cni/log/logger.go

```diff
@@ -30,13 +32,25 @@ func initZapLog(logFile string) *zap.Logger {
 	encoderConfig.EncodeTime = zapcore.ISO8601TimeEncoder
 	jsonEncoder := zapcore.NewJSONEncoder(encoderConfig)
 
-	core := zapcore.NewCore(jsonEncoder, logFileCNIWriter, zapcore.DebugLevel)
-	Logger := zap.New(core)
-	return Logger
+	textfilecore := zapcore.NewCore(jsonEncoder, logFileCNIWriter, loggingLevel)
+	Logger := zap.New(textfilecore, zap.AddCaller())
+	return Logger.With(zap.Int("pid", os.Getpid()))
+}
+
+func initZapLogWithETW(logFile string) *zap.Logger {
+	Logger := initZapLog(logFile)
+	etwcore, err := GetETWCore(ETWCNIEventName, loggingLevel)
+	if err != nil {
+		Logger.Error("Failed to attach ETW core to logger", zap.Error(err))
+		return Logger
+	}
+	teecore := zapcore.NewTee(Logger.Core(), etwcore)
+	Logger = zap.New(teecore, zap.AddCaller())
+	return Logger.With(zap.Int("pid", os.Getpid()))
```

Comment:
Finally, the whole `initZapLogWithETW` func goes away because it's not a concern of the root unplatformed logging code - which only needs to know that it is going to decorate the root logger core with some platform specific cores, and does _not_ need to know anything about the implementation of those platform specific things.

```suggestion
	textfilecore := zapcore.NewCore(jsonEncoder, logFileCNIWriter, loggingLevel)
	c, err := JoinPlatformCores(textfilecore, loggingLevel)
	if err != nil { 
		// should this be a fatal error? 
		// maybe we should simply fall back to the text core?
		c = textfilecore
	}
	return zap.New(c, zap.AddCaller()).With(zap.Int("pid", os.Getpid()))
}
```

pt. 3/3